Doc/graphql

JonnoKiwi opened the pull request:
Apart of progress plan "Explaining GraphQL vs RESTful"

JonnoKiwi committed:
GraphQL and Apollo Usage

JonnoKiwi committed:
server side

JonnoKiwi committed:
rework

JonnoKiwi committed:
upload diagram

johntimothybailey committed:
Merge branch 'master' into doc/graphql

johntimothybailey commented on readme.md:
```diff
@@ -315,7 +315,20 @@ In addition to `redux` --> `mobx-state-tree`, we've also transitioned to using `
+
+### Design/Architecture
+RESTful API uses HTTP verbs (Post, Get, Patch, etc...) in order to manipulate data. 
+Because of this, designing a RESTful system requires creating endpoints for each request such as: POST/user, GET/user, DELETE/user etc... 
```
@JonnoKiwi. This sentence doesn't seem correct. Verbs are an inherit property of HTTP. Your sentence makes it seem like you have to create separate endpoints for each Verb

JonnoKiwi replied:
I've updated my explanation

johntimothybailey commented on readme.md:
```diff
@@ -315,7 +315,20 @@ In addition to `redux` --> `mobx-state-tree`, we've also transitioned to using `
+RESTful API uses HTTP verbs (Post, Get, Patch, etc...) in order to manipulate data. 
+Because of this, designing a RESTful system requires creating endpoints for each request such as: POST/user, GET/user, DELETE/user etc... 
+
+GraphQL uses the HTTPS Post to send either Queries and Mutations to interact with a database. 
```
@JonnoKiwi needs a transition like "On the other hand, , GraphQL [...]" or "In contrast, , GraphQL [...]" or "To make it easier, GraphQL [...]"

johntimothybailey reviewed (changes requested):
Also, @JonnoKiwi would be good to give a pro and con for both.

JonnoKiwi committed:
update due to comments

JonnoKiwi committed:
Merge remote-tracking branch 'origin/doc/graphql' into doc/graphql

JonnoKiwi commented:
> Also, @JonnoKiwi would be good to give a pro and con for both.

I've added my conclusion that is listing some pros and cons

johntimothybailey reviewed (changes requested):
To be super clear the quote below is wrong
> “designing a RESTful system requires creating endpoints for each action” 

 - [POST] /users
- [ GET ] /users
- [ PATCH ] /users

are all one “endpoint” with three different verbs. Now you may have three different _**Controllers**_ (see Laravel for an example)

johntimothybailey commented:
@JonnoKiwi Also, let's get this in `docs/http-api-integration` with `docs/images/graphql-vs-restful.png`

JonnoKiwi committed:
relocate doc and image

JonnoKiwi commented:
> To be super clear the quote below is wrong
> 
> > “designing a RESTful system requires creating endpoints for each action”
> 
> * [POST] /users
> * [ GET ] /users
> * [ PATCH ] /users
> 
> are all one “endpoint” with three different verbs. Now you may have three different _**Controllers**_ (see Laravel for an example)

I've addressed this comment and changed the wording.

I've also relocated this writing and the image

johntimothybailey commented on readme.md:
```diff
@@ -315,7 +315,7 @@ In addition to `redux` --> `mobx-state-tree`, we've also transitioned to using `
 
 In Bowser, TypeScript is fully set up, so if you know TS, all you need to do is start coding!
 
-### Resources
```
@JonnoKiwi Why remove this?

JonnoKiwi replied:
Didnt mean to.

johntimothybailey reviewed (changes requested):
- `JSON:API`
- `GraphQL`
- Pure `RESTful`

- [ ] https://blog.runscope.com/posts/you-might-not-need-graphql
- [ ] https://nordicapis.com/how-does-jsonapi-compare-to-rest-and-graphql/
- [ ] https://dri.es/headless-cms-rest-vs-jsonapi-vs-graphql#:~:text=GraphQL%20and%20JSON%3AAPI%20were,typical%20inefficiency%20of%20REST%20APIs.&text=This%20makes%20it%20difficult%20to,lists%20of%20desired%20resource%20fields.
- [ ] https://www.apollographql.com/blog/graphql-vs-rest-5d425123e34b/

JonnoKiwi committed:
Update readme.md

JonnoKiwi committed:
Update http-api-intigration.md

this is not done, just need to save what I got

JonnoKiwi committed:
Update http-api-intigration.md

JonnoKiwi committed:
Update http-api-intigration.md

JonnoKiwi committed:
Update http-api-intigration.md

johntimothybailey commented on docs/http-api-intigration.md:
```diff
@@ -0,0 +1,65 @@
+## Helping out RESTful with JSON API or GraphQL
+
+### Design
+The design of RESTful API is to return resources using a Univerisal Resource Identifier. A client could make a requests such as `GET https//example.com/book/Hobbit`. This works in the simplest way to return all data from that resource even if all the fields aren't needed. If you needed specific information like `author` and `title`, you'd need to send multiple request: `GET /Hobbit/author`, `GET  ...
```
One of the core constraints* RESTful API  [.... about the URI]. However, there is not an explicit constraint for limiting or including additional fields in a resource response (e.g. techniques "Sparse Fields" and "Compound Documents" and "Pagination"). This leaves it up to the implementors to decide how or even **if** to implement these features. 

This is where [JSON:API](https://jsonapi.org/format/) and GraphQL (among a few others) provide clear constraints and features for the previously listed omissions RESTful API. 

While there are many features and constraints that JSON:API and GraphQL both provide, we will focus on the ones for reducing the size of resource responses and the number of requests need to get the data for a given application state. 

\* _According to _

JonnoKiwi replied:
Your wording is so good that I want to use most of it in the final product. Is that fine if I copy it?

JonnoKiwi committed:
Update http-api-intigration.md

JonnoKiwi committed:
Merge branch 'master' into doc/graphql

johntimothybailey committed:
Update http-api-intigration.md

johntimothybailey committed:
Update http-api-intigration.md

johntimothybailey committed:
Update http-api-intigration.md

johntimothybailey committed:
Update readme.md

JonnoKiwi committed:
doc: move MST and Sagasause documentation to state-machine-integration.md

JonnoKiwi committed:
doc: fixed type in filename

JonnoKiwi committed:
doc: fixed types and reworded Architecture section

johntimothybailey committed:
spelling and grammer corrections

johntimothybailey committed:
Merge branch 'master' into doc/graphql

johntimothybailey commented on docs/http-api-integration.md:
```diff
@@ -0,0 +1,106 @@
+## Architecture
+
+### RESTful
+RESTful API uses HTTP verbs (Post, Get, Patch, etc...) in order to manipulate data. 
```
I think this should be
> in order to retrieve and manipulate data

johntimothybailey commented on docs/http-api-integration.md:
```diff
@@ -0,0 +1,106 @@
+### RESTful
+RESTful API uses HTTP verbs (Post, Get, Patch, etc...) in order to manipulate data. 
+Because of this, designing a RESTful system requires creating controllers that enact actions (e.g. [POST] /user, [GET] /user/address, [PATCH] /user/age) 
+JSON:API does not deviate from the Controller-Model structure that RESTful uses. 
```
How so?

johntimothybailey commented on docs/http-api-integration.md:
```diff
@@ -0,0 +1,106 @@
+
+
+### GraphQL
+In GraphQL, Queries are sent with HTTP Get to retrieve resources from a database. When modifying data, it is NOT recommended to use HTTP Get. See documentation on Mutations at this [link](https://graphql.org/learn/queries/#mutations). 
```
I don't know how this sentence connects to the above or how it is about "Architecture" Perhaps you mean "Workflow" or something similar.

JonnoKiwi replied:
I don't think I add anything useful with this Architecture section. On top of that, the GraphQL diagram doesnt reflect how I describe the usage of GET and POST. I'm just gonna remove the whole section and end with "Final Thoughts"

johntimothybailey commented on docs/http-api-integration.md:
```diff
@@ -0,0 +1,106 @@
+### GraphQL
+**GraphQL** can also reduce the fields returned by specifying them in the GraphQL Query. Keep in mind, as its name implies GraphQL is predominantly a Query Language
+
+#### Sparse Fieldsets
```
I think that in the comparisons we should explain in GraphQL what is meant by `Sparse Fieldset` since GraphQL. This line just simply shows examples but really doesn't support the claim we made early on

johntimothybailey replied:
Quote the statement from the core team on their docs

Then explain something similar to 
> The `POST` verb and queries in the `body` enables complex queries that are not easily done (or at all) with `GET` verb and queries in the URI ([citation](https://graphql.github.io/learn/serving-over-http/)). We are not going into detail on the complex queries at this time, but you can read more about it in their documentation including [**Fragments**](https://graphql.github.io/learn/queries/#fragments)

johntimothybailey commented on docs/http-api-integration.md:
```diff
@@ -0,0 +1,106 @@
+GraphQL suggests various way to paginate returning data by requesting a set such as `first:10 offset:20` which is the equivalent to saying "second page of 10 items". 
+Another example is `first:5 offset:$itemId`, equivalent of saying "first 5 items after the ID of this item".
+https://graphql.org/learn/pagination/
+
```
Add a conclusion section (or suggestion)
> If you are starting out learning HTTP APIs, it may be easier to understand JSON:API. The ease of use with JSON:API is the reason why Hydrogen supports this as on option to GraphQL.

JonnoKiwi committed:
Update http-api-integration.md

johntimothybailey commented on docs/http-api-integration.md:
````diff
@@ -0,0 +1,112 @@
+https://graphql.org/learn/queries/#fields
+
+```
+GET /graphql
````
POST
_or_
GET with queries

johntimothybailey commented on docs/http-api-integration.md:
```diff
@@ -0,0 +1,112 @@
+
+### GraphQL
+**GraphQL** has features such as **Sparse Fields**, **Compound Documents**, **Pagination**, and **Caching**.
+
```
Something like

As stated by Paul _____ ([citation](https://blog.runscope.com/posts/you-might-not-need-graphql)) and ____ ([citation](https://apisyouwonthate.com/blog/understanding-rpc-rest-and-graphql)), GraphQL is more similar to a Remote Procedure Call API  than to a standard REST-style API like JSON:API.

JonnoKiwi commented on docs/http-api-integration.md:
```diff
@@ -0,0 +1,112 @@
+4. Uniform Interface
+5. Layered System
+6. Code-On-Demand
+
```
Jonno needs to state two definitions:
RPC: Remote Procedure Control
RESTful: Representational 

Having these stated will be helpful in showing the difference in behaviour how client and server interactions between JSON:API and GraphQL (In JSON:API, the client describes the action that it wants, while in GraphQL the client is telling the server what to do)

JonnoKiwi committed:
doc: rewording GraphQL features and removing Architecture section

JonnoKiwi committed:
Update http-api-integration.md

JonnoKiwi committed:
Update http-api-integration.md

JonnoKiwi committed:
Update http-api-integration.md

johntimothybailey committed:
Update http-api-integration.md

johntimothybailey committed:
Update http-api-integration.md
